espnet2/bin/enh_inference_streaming.py

```diff
@@ -0,0 +1,509 @@
+
+    # 3. Build data-iterator
+    loader = EnhancementTask.build_streaming_iterator(
+        data_path_and_name_and_type,
+        dtype=dtype,
+        batch_size=batch_size,
+        key_file=key_file,
+        num_workers=num_workers,
+        preprocess_fn=EnhancementTask.build_preprocess_fn(
+            separate_speech.enh_train_args, False
+        ),
+        collate_fn=EnhancementTask.build_collate_fn(
+            separate_speech.enh_train_args, False
+        ),
+        allow_variable_data_keys=allow_variable_data_keys,
+        inference=True,
+    )
+
+    # 4. Start dataset for-loop
+    output_dir = Path(output_dir).expanduser().resolve()
+    writers = []
+    for i in range(separate_speech.num_spk):
+        writers.append(
+            SoundScpWriter(f"{output_dir}/wavs/{i + 1}", f"{output_dir}/spk{i + 1}.scp")
+        )
+
+    import tqdm
+
+    for i, (keys, batch) in tqdm.tqdm(enumerate(loader)):
+        logging.info(f"[{i}] Enhancing {keys}")
+        assert isinstance(batch, dict), type(batch)
+        assert all(isinstance(s, str) for s in keys), keys
+        _bs = len(next(iter(batch.values())))
+        assert len(keys) == _bs, f"{len(keys)} != {_bs}"
+        batch = {k: v for k, v in batch.items() if not k.endswith("_lengths")}
+
+        speech = batch["speech_mix"]
+
+        # split continuous speech into small chunks to simulate streaming
+        speech_sim_chunks, rest_pad = split_audio(speech, frame_size, hop_size)
+        output_chunks = [[] for ii in range(separate_speech.num_spk)]
+
+        # the main loop for streaming processing
+        for chunk in speech_sim_chunks:
+            # process a single chunk
+            output = separate_speech(chunk)
+            for channel in range(separate_speech.num_spk):
+                # append processed chunks to ouput channels
+                output_chunks[channel].append(output[channel])
+
+        # reset separator states after processing
+        separate_speech.reset()
+
+        # merge chunks
+        waves = [
+            merge_audio(chunks, frame_size, hop_size, rest_pad)
+            for chunks in output_chunks
+        ]
+
+        waves = [
+            (w / abs(w).max(dim=1, keepdim=True)[0] * 0.9).cpu().numpy() for w in waves
+        ]
+
+        # waves = [w.cpu().numpy() for w in waves]
+
+        for spk, w in enumerate(waves):
+            for b in range(batch_size):
+                writers[spk][keys[b]] = fs, w[b]
+
+    for writer in writers:
+        writer.close()
+
+
+def merge_audio(chunks, frame_size, hop_size, rest):
+    # [chunk: (B, frame_size)]
+
+    num_chunks = len(chunks)
+    batch_size = chunks[0].shape[0]
+    audio_len = int(hop_size * num_chunks + frame_size - hop_size)
+
+    output = torch.zeros((batch_size, audio_len), dtype=chunks[0].dtype).to(
+        chunks[0].device
+    )
+
+    for i, chunk in enumerate(chunks):
+        output[:, i * hop_size : i * hop_size + frame_size] += chunk
```

Comment:
how's the energy of the overlapped region between two consecutive chunks? I feel the energy is doubled.

Reply:
I agree. The current code only works for some specific windows that always sums to 1 in the overlapped regions.

Please refer to the implementation in torchaudio: https://github.com/pytorch/audio/blob/release/0.5/torchaudio/functional.py#L185-L217

Reply:
Done! And now, all the consistency checks are passed.